agent/sidecar_service.go

```diff
@@ -114,9 +114,32 @@ func (a *Agent) sidecarServiceFromNodeService(ns *structs.NodeService, token str
 		}
 	}
 
+	port, err := a.sidecarPortFromServiceIDLocked(sidecar.Port, sidecar.CompoundServiceID())
+	if err != nil {
+		return nil, nil, "", err
+	}
+	sidecar.Port = port
+
+	// Setup checks
+	checks, err := ns.Connect.SidecarService.CheckTypes()
+	if err != nil {
+		return nil, nil, "", err
+	}
+	// Setup default check if none given
+	if len(checks) < 1 {
+		checks = sidecarDefaultChecks(ns.ID, sidecar.Proxy.LocalServiceAddress, sidecar.Port)
+	}
+
+	return sidecar, checks, token, nil
+}
+
+// sidecarPortFromServiceID is used to allocate a unique port for a sidecar proxy.
+// This is called immediately before registration to avoid value collisions. This function assumes the state lock is already held.
+func (a *Agent) sidecarPortFromServiceIDLocked(sidecarPort int, sidecarCompoundServiceID structs.ServiceID) (int, error) {
```

Comment:
IMO the `-Locked` suffix is most helpful when the locks exist near the functions being called. In this refactor I had to search the call stack to check that a state lock was being held somewhere so I think it might trip up other devs who look at this function and don't see a lock in scope.

Not sure if there's a clean solution (since it is worth calling out that we need a lock) but just calling it out.

Reply:
Signature-wise it's a little confusing to take `sidecarPort` as input and also return a port. Could we clarify in the docs what the input is for or perhaps mutate the sidecar directly?

Reply:
done